### PR TITLE
#5276 [Manager] Refactor the workspace deletion

### PR DESCRIPTION
**Pull request by eechih**
### Checklist
- [x] [Review Contribution Guidelines](https://oharastream.github.io/en/docs/master/contributing/)
- [x] Add Reviewers (see [Authors](https://github.com/oharastream/ohara#ohara-team))
- [x] Add Assignees (of course, you are the assignee by default!)
- [x] DON'T set a Milestone
- [x] DON'T add any labels (PR labels will be added automatically by Jenkins)
- [x] Link to an issue(You can do so by using the required by keyword: required by #5276 )
- [ ] Add unit tests (or please explain why this PR is not worth having new tests)
- [ ] Pass QA (Sometimes you may encounter an existing flaky test which would obstruct you from getting QA passed. If the failed tests are unrelated, you can add a comment in the PR thread)

### CODE REVIEW

File: ohara-manager/client/src/components/Workspace/Settings/DangerZone/DeleteWorkspace.js [hunk 1]
```diff
@@ -15,46 +15,118 @@
...
-      }}
-      steps={steps}
-    />
+    <Dialog
+      data-testid="delete-workspace"
+      fullWidth={isOpen}
+      maxWidth={'md'}
+      open={isOpen}
+    >
+      <DialogTitle>{'Delete Workspace 2'}</DialogTitle>
```
Comment (saivirtue) [inline ohara-manager/client/src/components/Workspace/Settings/DangerZone/DeleteWorkspace.js:115]: Delete Workspace 2 ?

File: ohara-manager/client/src/hooks/pipelineHooks.js [hunk 2]
```diff
@@ -135,6 +135,23 @@ export const useDeletePipelineAction = () => {
   );
 };
 
+export const useDeletePipelinesAction = () => {
+  const dispatch = useDispatch();
+  return useCallback(
+    (workspaceKey) =>
```
Comment (saivirtue) [inline ohara-manager/client/src/hooks/pipelineHooks.js:141]: Should we use the same argument `workspaceName` as `shabondiHooks`?

File: ohara-manager/client/src/observables/connectors.ts [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+  return defer(() => connectorApi.remove(key));
+}
+
+// Fetch and stop all connectors for this workspace
+export function fetchAndStopConnectors(
+  workspaceKey: ObjectKey,
+): Observable<ConnectorData[]> {
+  return fetchPipelines(workspaceKey).pipe(
+    mergeMap((pipelines) => {
+      if (isEmpty(pipelines)) return of([]);
```
Comment (saivirtue) [inline ohara-manager/client/src/observables/connectors.ts:115]: I will prefer to use [filter](https://rxjs-cn.github.io/learn-rxjs-operators/operators/filtering/filter.html) here :)

File: ohara-manager/client/src/components/Workspace/CreateWorkspace/CreateWorkspace.js [hunk 4]
```diff
@@ -65,19 +65,16 @@ export default () => {
         name: 'create zookeeper',
         action: () => createZookeeper(values?.zookeeper),
         revertAction: () => deleteZookeeper(values?.zookeeper?.name),
-        delay: 1000,
```
Comment (konekoya) [inline ohara-manager/client/src/components/Workspace/CreateWorkspace/CreateWorkspace.js:68]: 👍

File: ohara-manager/client/src/components/Workspace/Settings/DangerZone/DeleteWorkspace.js (hunk 1 above)
Comment (eechih) [reply ohara-manager/client/src/components/Workspace/Settings/DangerZone/DeleteWorkspace.js:115]: forgot to delete :(

File: ohara-manager/client/src/observables/connectors.ts (hunk 3 above)
Comment (eechih) [reply ohara-manager/client/src/observables/connectors.ts:115]: Good, I will try it :)

File: ohara-manager/client/src/const.ts [hunk 5]
```diff
@@ -125,3 +125,10 @@ export enum ACTIONS {
   START_BROKER = 'START_BROKER',
   START_ZOOKEEPER = 'START_ZOOKEEPER',
 }
+
+// retry every 2 seconds, up to 10 times
+export const RETRY_CONFIG = {
+  initialInterval: 2000, // 2 seconds
+  maxRetries: 10,
+  maxInterval: 2000, // 2 seconds
```
Comment (konekoya) [inline ohara-manager/client/src/const.ts:133]: How about _2 seconds interval between each retry_  :)

File: ohara-manager/client/src/hooks/shabondiHooks.js [hunk 6]
```diff
@@ -119,6 +139,26 @@ export const useDeleteShabondiAction = () => {
...
+  return useCallback(
+    (workspaceName) =>
+      new Promise((resolve, reject) =>
+        dispatch(
+          actions.deleteShabondis.trigger({
+            values: {
+              workspaceKey: { name: workspaceName, group: workspaceGroup },
+            },
+            resolve,
+            reject,
```
Comment (konekoya) [inline ohara-manager/client/src/hooks/shabondiHooks.js:154]: @eechih I remember we discussed the action args. The 1st arg is values (or params depending on the request type) and 2nd arg is an options object which can contain multiple items inside it. And now if we wrap an action in a promise. The order will be changed, right? I'm assuming: values (or params), options, resolve, reject?

File: ohara-manager/client/src/const.ts (hunk 5 above)
Comment (eechih) [reply ohara-manager/client/src/const.ts:133]: 預設的 backoffDelay 是 `exponentialBackoffDelay`，意味著每次 retry 後，將會把 interval 翻倍。例如 initialInterval 為 2s， 因此 intervals 為 2, 4, 8, 16, 32s, ....。而 maxInterval 則是限制每次的 interval 的長度，因此 intervals 為 2, 2, 2, 2, 2s, ....

File: ohara-manager/client/src/hooks/shabondiHooks.js (hunk 6 above)
Comment (eechih) [reply ohara-manager/client/src/hooks/shabondiHooks.js:154]: Yes, your assuming is right :)

Comment (eechih) [reply ohara-manager/client/src/hooks/shabondiHooks.js:154]: values (or params), options, resolve, reject 是放在 action payload 裡，沒有順序的問題。而 action parameters 則由每個 action 自訂。

File: ohara-manager/client/src/const.ts (hunk 5 above)
Comment (konekoya) [reply ohara-manager/client/src/const.ts:133]: Yep, so how about we change the comment from _2 seconds_ to _2 seconds interval between each retry_?

**Comment by saivirtue** [comment]
After testing in my local laptop, there are some small questions:
1. could we close the `Settings Dialog` after deleting workspace successfully? Since "Settings" page is behind the workspace
2. Is there an issue which will handle the unexpected network exceptions (ex: refresh page during deleting, connection-loss)? The deletion progress may stuck in following situation:
2.1. delete workspace
2.2. when the progress is at step 8 (Stop Broker), refresh page
2.3. delete workspace again
2.4. the progress will be failed at step 6 (Stop Worker): `java.util.NoSuchElementException`
3. The **Snackbar** will keep popup if we click the "RETRY" button more than 10 times

File: ohara-manager/client/src/hooks/pipelineHooks.js (hunk 2 above)
Comment (eechih) [reply ohara-manager/client/src/hooks/pipelineHooks.js:141]: 我思考由於 workspaceName 必須搭配 workspaceGroup，而 group 由 action 自己 handle，只開放 name 傳入，彈性只做一半。所以應該只有兩種選擇，一是完全開放彈性，或是不開放彈性。你覺得呢？ @saivirtue

Comment (saivirtue) [reply ohara-manager/client/src/hooks/pipelineHooks.js:141]: group現在UI上已有固定的規則，我想或許可以限制這個彈性?
目前我是沒有想到會有要由呼叫方改變group的情況...

File: ohara-manager/client/src/hooks/shabondiHooks.js (hunk 6 above)
Comment (konekoya) [reply ohara-manager/client/src/hooks/shabondiHooks.js:154]: > 沒有順序的問題

No problem. Didn't see it's an object :)

File: ohara-manager/client/src/const.ts (hunk 5 above)
Comment (eechih) [reply ohara-manager/client/src/const.ts:133]: No problem. Thanks @konekoya

File: ohara-manager/client/src/hooks/pipelineHooks.js (hunk 2 above)
Comment (eechih) [reply ohara-manager/client/src/hooks/pipelineHooks.js:141]: 我也覺得可以限制彈性。不過命名可能要清楚一點，例如 `useDeletePipelinesInWorkspaceAction`

File: ohara-manager/client/src/observables/connectors.ts (hunk 3 above)
Comment (eechih) [reply ohara-manager/client/src/observables/connectors.ts:115]: @saivirtue 這個 Case 不適合用 `filter`，因為 filter 之後會繼續執行 `forkJoin`。原本的邏輯會用 if 判斷，是為了避免 `forkJoin` 沒有任何 source 導致一直等不到任何 resolve 而卡住。

**Comment by eechih** [comment]
@saivirtue 

> 1. could we close the Settings Dialog after deleting workspace successfully? Since "Settings" page is behind the workspace

Sure, just fixed.

> 2. Is there an issue which will handle the unexpected network exceptions (ex: refresh page during deleting, connection-loss)? The deletion progress may stuck in following situation:
2.1. delete workspace
2.2. when the progress is at step 8 (Stop Broker), refresh page
2.3. delete workspace again
2.4. the progress will be failed at step 6 (Stop Worker): java.util.NoSuchElementException

Track this problem in #5321 

> 3. The Snackbar will keep popup if we click the "RETRY" button more than 10 times

will resolved by #5333

**Comment by eechih** [comment]
retry manager-e2e

**Comment by eechih** [comment]
retry manager-e2e

**Comment by eechih** [comment]
retry manager-e2e

**Comment by eechih** [comment]
retry manager-it

**Comment by eechih** [comment]
retry manager-it

**Comment by eechih** [comment]
retry manager-it

**Comment by eechih** [comment]
retry manager-it

**Pull request merged**